lib/ecto/type.ex

```diff
@@ -88,15 +88,18 @@ defmodule Ecto.Type do
     end
   end
 
-  @typedoc "An Ecto type, primitive or custom."
-  @type t :: primitive | custom
+  @typedoc "An Ecto type: primitive, custom, or parameterized."
+  @type t :: primitive | custom | parameterized
 
   @typedoc "Primitive Ecto types (handled by Ecto)."
   @type primitive :: base | composite
 
   @typedoc "Custom types are represented by user-defined modules."
   @type custom :: module
 
+  @typedoc "Parameterized custom types are represented by user-defined modules and options."
+  @type parameterized :: {module, keyword()} | {:parameterized, module, keyword()}
```

Comment:
I am not a big fan of the double representation but I am not sure if we have other alternatives. I guess one option is to make the type a `map` or a `struct`? This way you can dispatch to it by calling `type.__struct__.dump()` and so on. WDYT?

Reply:
To be a bit more precise, we would declare them for now as `field :foo, %Ecto.Enum{values: [:foo, :bar]}`. It is not the best API in the world but we can always add sugar later.

Reply:
The {module, keyword()} representation only appears in the __field__ call, and Dialyzer was complaining in check_field_type! because primitive? takes a Type.t. I can catch and convert those before check_field_type! and remove the {module, keyword()} version from type.

Reply:
Ok I think I understand what you mean: It would be best if the internally tracked "type" was simply the module that implements parameterized, with the opts stored inside them, instead of explicitly internally handling {:parameterized, type, opts}. This would be similar to how Custom types are handled, but would require determining if a module is custom or parameterized at a number of stages in the code.

Given what I have seen so far on this, I see the benefit of that approach but I still would lean towards maintaining the explicit :parameterized tuple internally because it makes branching on the logic that needs to be handled different much cleaner, with less guess work and less error prone. The lines between a custom type and a parameterized type would get blurry and we would have to pick some test to test if the type is a custom or parameterized type everywhere that we currently assume it is custom. Alternatively, with the tuple method, we know for a fact that the user intended the type to be parameterized from the start in the __field__ call, so carrying that over in a :parameterized tuple to me feels like the safer way, but let me know which one you would like and I will adjust accordingly.

Reply:
> The {module, keyword()} representation only appears in the field call, and Dialyzer was complaining in check_field_type! because primitive? takes a Type.t. I can catch and convert those before check_field_type! and remove the {module, keyword()} version from type.

Yes, let's go this route then. Please ignore my second comment. :)

Reply:
> The {module, keyword()} representation only appears in the field call, and Dialyzer was complaining in check_field_type! because primitive? takes a Type.t. I can catch and convert those before check_field_type! and remove the {module, keyword()} version from type.

Yes, let's go this route then. Please ignore my second comment. :)